File: DSA_chaicode/two_pointer/dsa.py
def twoSum(arr: list[int], target: int) -> list[int]:
    left: int = 0
    right: int = (len(arr)) - 1

    while left < right:
        if arr[left] + arr[right] == target:
            return [left+1, right+1]
        elif arr[left] + arr[right] < target:
            left += 1
        else: right -= 1
    return [-1, -1]

def validPalindrome(s: str) -> bool:
    left: int = 0
    right: int = (len(s)) - 1

    while left < right:
        while left < right and not s[left].isalnum():
            left += 1
        while left < right and not s[right].isalnum():
            right -= 1
        if s[left].lower() != s[right].lower():
            return False

        left += 1;
        right -= 1;

    return True

def main():
    # TwoSum
    arr1 = [2, 3, 5, 6, 10]
    print(twoSum(arr1, 5))

    #palindrome
    s = "A man, a plan, a canal: Panama"
    print(validPalindrome(s))

main()

File: DSA_chaicode/two_pointer/dsa.ts
function twoSum(arr: number[], target: number): number[] {

    let left: number = 0;
    let right: number = arr.length - 1

    while(left < right){
        let sum = arr[left] + arr[right];
        if(sum === target) return [left + 1, right + 1];
        else if(sum < target) left++;
        else right--;
    }
    return [-1, -1]
}

function validPalindrome(s: string): boolean{
    const cleaned = s.replace(/[^a-zA-Z0-9]/g, "").toLowerCase();
    let left = 0;
    let right = cleaned.length - 1;
    while(left < right){
        if(cleaned[left] !== cleaned[right]) return false;
        left++;
        right--;
    }
    return true;
}

function main(){
    // TwoSum
    const arr1 = [2, 3, 5, 6, 10];
    console.log(twoSum(arr1, 5))

    //palindrome
    const s = "A man, a plan, a canal: Panama"
    console.log(validPalindrome(s))
}

main()